lib/grandpa/vote_message.go

```diff
@@ -23,58 +22,169 @@ type networkVoteMessage struct {
 }
 
 // receiveVoteMessages receives messages from the in channel until a grandpa round finishes.
-func (s *Service) receiveVoteMessages(ctx context.Context) {
-	for {
-		select {
-		case msg, ok := <-s.in:
-			if !ok {
+func (s *Service) receiveVoteMessages(cancel <-chan struct{}) (determinePrecommitCh, finalizableCh chan struct{}) {
+	determinePrecommitCh = make(chan struct{})
+	finalizableCh = make(chan struct{})
+
+	go func() {
+		defer func() {
+			close(determinePrecommitCh)
+			if finalizableCh != nil {
+				close(finalizableCh)
+			}
+		}()
+
+		logger.Debug("receiving pre-vote messages...")
+		threshold := s.state.threshold()
+
+		// we should determine the precommit only when we reach the
+		// prevotes threshold otherwise we should wait for them
+		prevotesThresholdReached := false
+
+		for {
+			select {
+			case commit, ok := <-s.receivedCommit:
+				if !ok {
+					return
+				}
+
+				if commit.Round == s.state.round {
+					close(finalizableCh)
+					finalizableCh = nil
+					return
+				}
+
+			case msg, ok := <-s.in:
+				if !ok {
+					return
+				}
+
+				if msg == nil || msg.msg == nil {
+					continue
+				}
+
+				vm := msg.msg
+				logger.Debugf("received vote message %v from %s", msg.msg, msg.from)
+				s.sendTelemetryVoteMessage(msg.msg)
+
+				v, err := s.validateVoteMessage(msg.from, vm)
+				if err != nil {
+					logger.Debugf("failed to validate vote message %v: %s", vm, err)
+					continue
+				}
+
+				logger.Debugf(
+					"validated vote message %v from %s, round %d, subround %d, "+
+						"prevote count %d, precommit count %d, votes needed %d",
+					v, vm.Message.AuthorityID, vm.Round, vm.Message.Stage,
+					s.lenVotes(prevote), s.lenVotes(precommit), s.state.threshold()+1)
+
+				// when a given vote is validated we should check
+				// if we have reached the prevotes threshold
+				prevotesThreshold := s.lenVotes(prevote) - 1
+				if !prevotesThresholdReached && prevotesThreshold >= int(threshold) {
+					prevotesThresholdReached = true
+					determinePrecommitCh <- struct{}{}
```

Comment:
Explaining why I don't close the round here:

In the `playGrandpaRound` we have this blocking loop where we wait for one of 3 channels:
1. context closure then we exit the function
2. Enough pre-votes then we should determine and send our precommit
3. Round is finalizable means we reach enough amount or pre-commits or we received a valid commit message that finalizes a block for the current round

```go
for {
  select {
    case <-s.ctx.Done():
      return nil
    case <-determinePrecommit:
      ...
      go s.sendPrecommitMessage(pcm, cancel)
    case <-finalizable:
      // send the commit message and jump into the next round		
  }
}
```

Specifically, the second case is controlled here where we check if we reach the amount of pre-votes needed to determine our pre-commit, and the point is: if we close the channel instead of just passing and empty `struct{}` then in our loop we will always trigger the second channel since closed channels are always good to read unless you check `_, ok := <-channel` the closed property

```go
prevotesThreshold := s.lenVotes(prevote) - 1
if !prevotesThresholdReached && prevotesThreshold >= int(threshold) {
  prevotesThresholdReached = true
  // determinePrecommitCh <- struct{}
  close(determinePrecommitCh)
}
```